File: android/client/src/com/qorporation/msgs/client/android/account/AccountInitializer.java
package com.qorporation.msgs.client.android.account;

import org.json.JSONObject;

import com.qorporation.msgs.client.android.networking.HttpRequest;
import com.qorporation.msgs.client.android.networking.HttpRequestDelegate;
import com.qorporation.msgs.client.android.store.AccountStore;

public class AccountInitializer implements HttpRequestDelegate {

	private AccountInitializerDelegate delegate = null;
	private HttpRequest request = null;

	public AccountInitializer(AccountInitializerDelegate delegate) {
		this.delegate = delegate;
		this.request = new HttpRequest(String.format(
				HttpRequest.AUTH_REQUEST_URL_FORMAT,
				AccountStore.getInstance().getDeviceIdent(),
				AccountStore.getInstance().getDeviceVerifier(),
				AccountStore.getInstance().getAuthToken()), this);
	}
	
	public void start() {
		this.request.send();
	}
	
	public void onRequestFailure(HttpRequest request) {
		this.delegate.onAccountInitializationFailed();
	}

	public void onRequestSending(int sent, int total, HttpRequest request) {
	}

	public void onRequestReceiving(int received, int total, HttpRequest request) {
	}

	public void onRequestComplete(String response, HttpRequest request) {
		try {
			JSONObject json = new JSONObject(response);
			String status = json.getString("status");
			if (status != null && status.equals("ok")) {
				String userID = json.getString("user_id");
				String userName = json.getString("user_name");
				String userEmail = json.getString("user_email");
				String userEmailPending = json.getString("user_email_pending");
				String userPhone = json.getString("user_phone");
				String userPhonePending = json.getString("user_phone_pending");
				String authToken = json.getString("auth_token");
				String authSecret = json.getString("auth_secret");
				
				AccountStore.getInstance().setUserID(userID);
				AccountStore.getInstance().setUserName(userName);
				AccountStore.getInstance().setUserEmail(userEmail);
				AccountStore.getInstance().setUserEmailPending(userEmailPending);
				AccountStore.getInstance().setUserPhone(userPhone);
				AccountStore.getInstance().setUserPhonePending(userPhonePending);
				AccountStore.getInstance().setAuthToken(authToken);
				AccountStore.getInstance().setAuthSecret(authSecret);
				
				this.delegate.onAccountInitialized(userID, userEmail, authToken, authSecret);
			} else {
				this.delegate.onAccountInitializationFailed();
			}
		} catch (Exception e) {
			this.delegate.onAccountInitializationFailed();
		}
	}

}

File: iphone/msgs/Classes/AccountInitializer.m
//
//  AccountInitializer.m
//  msgs
//
//  Created by Lewis Zimmerman on 11-01-15.
//  Copyright 2011 Qorporation. All rights reserved.
//

#import "AccountInitializer.h"
#import "Atlas.h"

#import <Three20Core/Three20Core.h>
#import <Three20Core/TTCorePreprocessorMacros.h>
#import <Three20Network/TTGlobalNetwork.h>
#import <Three20Network/TTURLRequest.h>
#import <extThree20JSON/extThree20JSON.h>

@interface AccountInitializer ()

- (void)load;

@end

@implementation AccountInitializer

///////////////////////////////////////////////////////////////////////////////////////////////////
- (void)dealloc {
	TT_RELEASE_SAFELY(_delegate);
	
	[super dealloc];
}

///////////////////////////////////////////////////////////////////////////////////////////////////
- (id)initWithDelegate:(id<AccountInitializerDelegate>)delegate {
	_delegate = delegate;
	_store = [AccountStore get];	
	[self load];
	
	return self;	
}

///////////////////////////////////////////////////////////////////////////////////////////////////
+ (id)start:(id<AccountInitializerDelegate>)delegate {
	return [[AccountInitializer alloc] initWithDelegate:delegate];
}

///////////////////////////////////////////////////////////////////////////////////////////////////
- (void)load {
	NSString* deviceIdent = [_store getDeviceIdent];
	NSString* deviceVerifier = [_store getDeviceVerifier];
	NSString* authToken = [_store getAuthToken];
	
	NSString* url = [NSString stringWithFormat:kAuthRequestURLFormat, deviceIdent, deviceVerifier, authToken ? authToken : @""];
	
	TTURLRequest* request = [TTURLRequest requestWithURL:url delegate:self];
	TTURLJSONResponse* response = [[TTURLJSONResponse alloc] init];
	request.response = response;
	TT_RELEASE_SAFELY(response);
	
	[request send];
}

///////////////////////////////////////////////////////////////////////////////////////////////////
- (void)requestDidFinishLoad:(TTURLRequest*)request {
	TTURLJSONResponse* response = request.response;
	TTDASSERT([response.rootObject isKindOfClass:[NSDictionary class]]);
	
	NSDictionary* info = response.rootObject;
	TTDASSERT([[info objectForKey:@"status"] isKindOfClass:[NSString class]]);
	
	NSString* status = [info objectForKey:@"status"];
	if ([status isEqualToString:@"ok"]) {
		TTDASSERT([[info objectForKey:@"user_id"] isKindOfClass:[NSString class]]);
		TTDASSERT([[info objectForKey:@"user_email"] isKindOfClass:[NSString class]]);
		TTDASSERT([[info objectForKey:@"auth_token"] isKindOfClass:[NSString class]]);
		TTDASSERT([[info objectForKey:@"auth_secret"] isKindOfClass:[NSString class]]);
		
		NSString* userID = [info objectForKey:@"user_id"];
		NSString* userEmail = [info objectForKey:@"user_email"];
		NSString* authToken = [info objectForKey:@"auth_token"];
		NSString* authSecret = [info objectForKey:@"auth_secret"];
		
		[_store setUserID:userID];
		[_store setUserEmail:userEmail];
		[_store setAuthToken:authToken];
		[_store setAuthSecret:authSecret];
		
		[_delegate accountInitialized:userID userEmail:userEmail authToken:authToken authSecret:authSecret];
	} else {
		[_delegate accountInitializationFailed];
	}
}

@end
